src/no_std.rs

```diff
@@ -270,9 +309,17 @@ impl<T> Drop for ListGuard<'_, T> {
         drop(list);
 
         // Wakeup all tasks.
-        for task in tasks {
+        for task in self.tasks.drain(..) {
             task.wake();
         }
+
+        // There is a deadlock where a node is pushed the end of the queue after we've finished
+        // process_nodes() but before we've finished dropping the lock. This can lead to systems
+        // halting. Therefore check before we finish dropping if there is anything left in the
+        // queue, and if so, lock it again and force a queue update.
+        if !self.inner.list.queue.is_empty() {
+            self.inner.queue_update();
```

Comment:
imo this should be implemented in a way that it won't recursively call the destructor; instead, it should do so iteratively (e.g. by replacing itself and looping until the queue is empty)